### PR TITLE
[ADD][9.0] New module stock_picking_rate

### PR DESCRIPTION
**Pull request by lasley**
This module introduces a concept of rate quotes and purchasing on Stock Pickings. This is one of the modules yielded from the discussion in OCA/carrier-delivery#90.
# Stock Picking Rate

This module adds the concept of delivery rate quotes & purchasing
on Stock Pickings

This is meant to be combined with an external connector with a carrier
supplier in order to bring in rate quotes, although they can be created &
edited manually if your workflow calls for it.
# Usage

Dispatch rates can be added in Stock Picking
## Purchase Rate

The Purchase Rate wizard can be accessed by clicking the green check next
to the rate quote in a `stock.picking`.

Completing this wizard will create `purchase.order` for your rate.
If using a connector, this is the point in which a shipment purchase would
be triggered.

### CODE REVIEW

**Review by rousseldenis — APPROVED** [review]
:+1: Tested on runbot. Seems great to use!

**Comment by pedrobaeza** [comment]
Stock picking dispatch has been renamed to stock batch picking I'm afraid...

**Comment by lasley** [comment]
Alright I'm just going to drop the `dispatch` suffix then, because the naming doesn't work with `stock_batch_picking`. It's not actually dependent, I was just trying to make the namespaces match

@t3ddftw - Take note of this change & please update in dependent connector

File: stock_picking_rate/models/stock_picking_rate.py
Comment (susport) [inline stock_picking_rate/models/stock_picking_rate.py]: Order by name the import

File: stock_picking_rate/tests/test_stock_picking_rate_purchase.py
Comment (woodbrettm) [inline stock_picking_rate/tests/test_stock_picking_rate_purchase.py:25]: @lasley when you have a chance can you take a look at this? It's returning an empty array, which is causing another test to fail as well.

Comment (lasley) [reply stock_picking_rate/tests/test_stock_picking_rate_purchase.py:25]: cc @tedsalmon

Comment (lasley) [reply stock_picking_rate/tests/test_stock_picking_rate_purchase.py:25]: Also this shoudl be `rec_id.rate_ids.ids`. But it would still be empty

**Comment by woodbrettm** [comment]
Woot the build is passing locally now. Although I think Travis is failing from something unrelated?

**Comment by lasley** [comment]
Yes as long as OCA/OCB build passes, you can assume odoo/odoo error is - https://odoo-community.org/groups/contributors-15/contributors-49663?mode=thread&date_begin=&date_end=

**Comment by woodbrettm** [comment]
Updated

File: stock_picking_rate/models/stock_picking_rate.py
Comment (obulkin) [inline stock_picking_rate/models/stock_picking_rate.py]: Seems like there should be a `depends` here.

File: stock_picking_rate/tests/test_stock_picking_rate_purchase.py
Comment (obulkin) [inline stock_picking_rate/tests/test_stock_picking_rate_purchase.py]: This method is not getting used anywhere. Also, the first line seems wrong. As it stands, if the call includes an argument for `vals`, that argument won't be used and `vals` will simply be set to an empty dictionary.

File: stock_picking_rate/models/stock_picking_rate.py
Comment (obulkin) [inline stock_picking_rate/models/stock_picking_rate.py:110]: Seems like overkill to wrap this up in a method. Is this done so that it can be overloaded later?

**Review by obulkin — CHANGES_REQUESTED** [review]
I'm seeing a few spots that may need changes, but generally, it LGTM.

File: stock_picking_rate/models/stock_picking_rate.py
Comment (lasley) [reply stock_picking_rate/models/stock_picking_rate.py:110]: Exactly, it's for overload. We now have a lot of these in v10 [API interface](https://github.com/odoo/odoo/blob/10.0/addons/delivery/models/delivery_carrier.py#L142)

File: stock_picking_rate/tests/test_stock_picking_rate_purchase.py
Comment (lasley) [reply stock_picking_rate/tests/test_stock_picking_rate_purchase.py]: Remove if not used

**Comment by woodbrettm** [comment]
@obulkin Thanks! Updated.

**Review by tedsalmon — APPROVED** [review]
👍  LGTM! There was an issue with files having the executable bit set, but I took care of that :)

**Review by obulkin — APPROVED** [review]
Looks like my comments were addressed, so this should be good to go now. 👍

**Comment by lasley** [comment]
@OCA/logistics-maintainers - Can a PSC please review this one? It's been sitting for a while & I would like to see merged so we can start forward port.

File: stock_picking_rate/models/stock_picking_rate.py
Comment (pedrobaeza) [inline stock_picking_rate/models/stock_picking_rate.py:82]: Isn't better to just write the date on the state change?

File: stock_picking_rate/models/stock_picking_rate.py
Comment (pedrobaeza) [inline stock_picking_rate/models/stock_picking_rate.py:56]: s/Est/Est.

File: stock_picking_rate/models/stock_picking_rate.py
Comment (pedrobaeza) [inline stock_picking_rate/models/stock_picking_rate.py:111]: Here is where you should write the date

File: stock_picking_rate/models/stock_picking_rate.py
Comment (pedrobaeza) [inline stock_picking_rate/models/stock_picking_rate.py:118]: You can compare directly records without `.id`

File: stock_picking_rate/tests/test_stock_picking_rate_purchase.py
Comment (pedrobaeza) [reply stock_picking_rate/tests/test_stock_picking_rate_purchase.py:25]: Please change this

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
The module name misses the main concept for this: delivery. The rate is for delivery, not for other thing, so the module should be called __stock_picking_delivery_rate__ IMO, and it should be hosted not here, but on https://github.com/oca/delivery-carrier

**Comment by lasley** [comment]
@bmw95 - please see to Pedro's comments & move this PR to the correct repo when done

@tedsalmon - please take note for your dependent module(s)

**Comment by lasley** [comment]
Closing in favor of OCA/delivery-carrier#119

**Pull request closed without merging**
